thrust/examples/bounding_box.cu

```diff
@@ -68,31 +66,38 @@ struct bbox_reduction
   }
 };
 
+// functor to generate random points
+struct random_point_generator
+{
+  mutable thrust::default_random_engine rng;
+  mutable thrust::uniform_real_distribution<float> u01;
+
+  random_point_generator()
+      : rng(0)
+      , u01(0.0f, 1.0f)
+  {}
+
+  __host__ __device__ point2d operator()() const
+  {
+    return point2d(u01(rng), u01(rng));
+  }
+};
+
 int main()
 {
   const size_t N = 40;
-  thrust::default_random_engine rng;
-  thrust::uniform_real_distribution<float> u01(0.0f, 1.0f);
 
   // allocate storage for points
   thrust::device_vector<point2d> points(N);
 
   // generate some random points in the unit square
-  for (size_t i = 0; i < N; i++)
-  {
-    float x   = u01(rng);
-    float y   = u01(rng);
-    points[i] = point2d(x, y);
-  }
+  thrust::generate(points.begin(), points.end(), random_point_generator());
```

Comment:
Important: I don't think this is legal. First, `generate` can copy the passed function at will, so it will duplicate the RNG state. Furthermore, where it does not copy the passed function, it can call the function in parallel and that will lead to a race condition. Please revert this portion.